src/proxy/chain.js

```diff
@@ -43,27 +46,48 @@ const chain = async (req) => {
   return action;
 };
 
+/**
+ * The plugin loader used for the Git Proxy chain.
+ *
+ * @type {import('../plugin').PluginLoader}
+ */
+let chainPluginLoader;
+
 const getChain = async (action) => {
-  if (action.type === 'pull') return [proc.push.checkRepoInAuthorisedList];
-  if (action.type === 'push') {
-    // insert loaded plugins as actions
-    // this probably isn't the place to insert these functions
-    const loader = await plugin.defaultLoader;
-    const pluginActions = loader.plugins;
-    if (!pluginsLoaded && pluginActions.length > 0) {
-      console.log(`Found ${pluginActions.length}, inserting into proxy chain`);
-      for (const pluginAction of pluginActions) {
-        if (pluginAction instanceof plugin.ActionPlugin) {
-          console.log(`Inserting plugin ${pluginAction} into chain`);
-          // insert custom functions after parsePush but before other actions
-          pushActionChain.splice(1, 0, pluginAction.exec);
-        }
-      }
-      pluginsLoaded = true;
+  if (chainPluginLoader === undefined) {
+    console.error('Plugin loader was not initialized! Skipping any plugins...');
+    pluginsLoaded = true;
+  }
+  if (!pluginsLoaded) {
+    console.log(`Inserting loaded plugins (${chainPluginLoader.pushPlugins.length} push, ${chainPluginLoader.pullPlugins.length} pull) into proxy chains`);
+    for (const pluginObj of chainPluginLoader.pushPlugins) {
+      console.log(`Inserting push plugin ${pluginObj.constructor.name} into chain`);
+      // insert custom functions after parsePush but before other actions
+      pushActionChain.splice(1, 0, pluginObj.exec);
```

Comment:
@coopernetes - very clean so far ❤️ Can we implement order of execution with the custom property on the plugin? Not required for this to be merged, but I'll open an issue once we merge if not. Great job, Thomas!

Reply:
Let's make that a separate enhancement. The core of this PR is to address the documentation definiciencies with some usability fixes that we encountered on our internal use of plugins.

Reply:
This is more than perfect, thank you 🍰